blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+refresh:537, AbstractApplicationContext (org.springframework.context.support)
+refresh:122, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+refresh:693, SpringApplication (org.springframework.boot)
+refreshContext:360, SpringApplication (org.springframework.boot)
+run:303, SpringApplication (org.springframework.boot)
+run:1118, SpringApplication (org.springframework.boot)
+run:1107, SpringApplication (org.springframework.boot)
+main:35, DubboConsumerDemo (com.alibaba.boot.dubbo.demo.consumer.bootstrap)
+```
+
+It can be seen that during the startup process of the Spring-boot application, the above method is executed since the Tomcat exposes the HTTP service by default, and all threads started by Tomcat are daemon threads by default, such as the Acceptor of the listening request, working thread pool etc.The JVM will also exit after the startup is completed if there is no control here. Therefore, it is necessary to explicitly start a thread and continue to wait under certain conditions. We can avoid thread exit by this way.
+
+Let's dig deeper. In Tomcat's `this.tomcat.getServer().await()` method, how the thread doesn‘t exit. The irrelevant code is removed here for the convenience of reading.
+
+```java
+public void await() {
+        // ...
+        if( port==-1 ) {
+            try {
+                awaitThread = Thread.currentThread();
+                while(!stopAwait) {
+                    try {
+                        Thread.sleep( 10000 );
+                    } catch( InterruptedException ex ) {
+                        // continue and check the flag
+                    }
+                }
+            } finally {
+                awaitThread = null;
+            }
+            return;
+        }
+        // ...
+    }
+```
+
+In the await method, the current thread checks the variable `stopAwait` every 10 seconds in a while loop. It is a `volatile` variable that is used to ensure that the current thread can see the change immediately after being modified by another thread. If there is no change, it will always be in the while loop. This is why the thread does not exit, which is why the entire Spring-boot application doesn't exit.
+
+Because Spring-boot application enable 2 port that 8080 and 8081(management port) at the same time. Two Tomcat are startup actually, so there are two threads that `container-0` and `container-1`.
+
+Then, let's see how this Spring-boot application exits.
+
+### The analysis of DubboConsumer exit
+
+As mentioned in the previous description, there is a thread that checks the variable `stopAwait` continuously. Then we can think that should be a thread to modify `stopAwait` at Stop, it can break the while loop. So, who is modifying this variable?
+
+By analyzing the source code, you can see that there is only one way to modify `stopAwait`, in `org.apache.catalina.core.StandardServer#stopAwait`, we add a breakpoint here to see who is calling.
```

Comment:
... we can see that there is only one method that modifies `stopAwait` : `org.....`. To figure out who is calling this method, we add a breakpoint here.